### DISCUSSION TITLE
select subquery inside returning

### DISCUSSION THREAD

**Discussion by cpoppema** [sqlalchemy/sqlalchemy#11665] [Usage Questions] [answered]
I'm trying to write queries in sqlalchemy such as highlighted in this SO post: https://stackoverflow.com/questions/7923237/return-pre-update-column-values-using-sql-only

```
UPDATE tbl x
SET    tbl_id = 24
     , name = 'New Gal'
FROM  (SELECT tbl_id, name FROM tbl WHERE tbl_id = 4 FOR UPDATE) y 
WHERE  x.tbl_id = y.tbl_id
RETURNING y.tbl_id AS old_id, y.name AS old_name
        , x.tbl_id          , x.name;
```

or

```
UPDATE users SET email = 'new@gmail.com' WHERE id = 1
RETURNING (SELECT email FROM users WHERE id = 1);
```

But as part of an insert..on conflict.. do update - if that part matters.

I have tried writing a .subquery(), but I either get an explicit error: `missing FROM-clause entry for table`, or the built query returns all columns from the subquery-selected table instead of what I try to specify.

My attempts
```
        insert_stmt = insert(MyModel).values(
            some_id=id,
            some_value=value,
            force_publish=False,
        )

        do_update_stmt = insert_stmt.on_conflict_do_update(
            index_elements=["some_id"],
            set_={
                "some_value": insert_stmt.excluded.some_value,
                "force_publish": False,
            },
        )

        force_publish_stmt = (
            select(MyModel.force_publish)
            .filter(MyModel.some_id == id)
            .subquery()
        )

        do_update_stmt = do_update_stmt.returning(
            MyModel.some_value,
            force_publish_stmt,
        )
```

and

```

        insert_stmt = insert(MyModel).values(
            some_id=id,
            some_value=value,
            force_publish=False,
        )

        do_update_stmt = insert_stmt.on_conflict_do_update(
            index_elements=["some_id"],
            set_={
                "some_value": insert_stmt.excluded.some_value,
                "force_publish": False,
            },
        )

        force_publish_stmt = aliased(
            MyModel,
            select(MyModel.force_publish)
            .filter(MyModel.some_id == id)
            .subquery()
        )

        do_update_stmt = do_update_stmt.returning(
            MyModel.some_value,
            force_publish_stmt,
        )
```

using aliased from sqlalchemy.orm to try to address the error: ` missing FROM-clause entry for table "anon_1"`, however this version simply puts all the table columns in the `RETURNING` instead of only the columns specified.

Basically an external action can set the force_publish boolean which an insert or update should reset - but I'd like to know if the flag was set. With high volume concurrent writes I want to avoid doing a select beforehand and these suggestions from stackoverflow to do a subquery select in the returning seemed to do what I want.

**Comment by zzzeek** [comment]
can you provide a sample model in context and a more complete example of what SQL you are looking to render, as well as what database you are targerting (postgresql?)  thanks

**Reply by zzzeek** [reply]
also the use of aliased() there so far does not seem warranted but im not sure what kind of rows you want to get back

**Reply by cpoppema** [reply]
Indeed, targeting postgresql. A more complete example of the desired/expected SQL would be something like this (based on the stackoverflow suggestions):

```sql
INSERT INTO account (id, valid_until, last_status, force_publish)
VALUES ($1::BIGINT,
        $2::TIMESTAMP WITH TIME ZONE,
        $3::VARCHAR,
        $4::BOOLEAN)
ON CONFLICT (account_id) DO UPDATE
SET
    valid_until = excluded.valid_until,
    last_status = CASE
      WHEN (account.valid_until < $5::TIMESTAMP WITH TIME ZONE)
      THEN $6::VARCHAR
      ELSE $7::VARCHAR
    END,
    force_publish = $8::BOOLEAN
RETURNING
    account.valid_until,
    account.last_status,
    (SELECT old_account.force_publish FROM account AS old_account WHERE id = id);
```

I'll work on a more complete example in the meantime.

**Reply by zzzeek** [reply]
edit: the "WHERE id = id" part is not valid, so you have to figure out how you want to place an "id" there, sending a specific value in is best

OK, here is a complete demonstration of that SQL, I think the missing link here is that if you are looking to return a SELECT subquery as a column you need to use `scalar_subquery()`, try this complete program

```py
from __future__ import annotations

import datetime

from sqlalchemy import case
from sqlalchemy import select
from sqlalchemy.dialects import postgresql
from sqlalchemy.dialects.postgresql import insert
from sqlalchemy.orm import aliased
from sqlalchemy.orm import DeclarativeBase
from sqlalchemy.orm import Mapped
from sqlalchemy.orm import mapped_column


class Base(DeclarativeBase):
    pass


class Account(Base):
    __tablename__ = "account"

    id: Mapped[int] = mapped_column(primary_key=True)
    account_id: Mapped[int]
    valid_until: Mapped[datetime.datetime]
    last_status: Mapped[str]
    force_publish: Mapped[bool]


stmt = insert(Account).values(
    id=1,
    valid_until=datetime.datetime.today() + datetime.timedelta(days=5),
    last_status="some status",
    force_publish=False,
)

old_account = aliased(Account)

stmt = stmt.on_conflict_do_update(
    index_elements=["account_id"],
    set_=dict(
        valid_until=stmt.excluded.valid_until,
        last_status=case(
            (
                Account.valid_until < datetime.datetime.today(),
                "some other status",
            ),
            else_="some status",
        ),
        force_publish=False,
    ),
).returning(
    Account.valid_until,
    Account.last_status,
    select(old_account.force_publish)
    .where(old_account.id == Account.id)  # this is not supported since there's no enclosing SELECT
    .scalar_subquery()
    .label("old_force_publish"),
)

print(stmt.compile(dialect=postgresql.dialect()))

# prints:
"""
INSERT INTO account
    (id, valid_until, last_status, force_publish)
    VALUES (%(id)s, %(valid_until)s, %(last_status)s, %(force_publish)s)
    ON CONFLICT (account_id) DO UPDATE
    SET
        valid_until = excluded.valid_until,
        last_status = CASE WHEN (account.valid_until < %(valid_until_1)s) THEN %(param_1)s ELSE %(param_2)s END,
        force_publish = %(param_3)s
    RETURNING
        account.valid_until,
        account.last_status,
        (SELECT account_1.force_publish
        FROM account AS account_1, account    -- this part is not what you want
        WHERE account_1.id = account.id) AS old_force_publish

"""

# requested SQL
"""
INSERT INTO account (id, valid_until, last_status, force_publish)
VALUES ($1::BIGINT,
        $2::TIMESTAMP WITH TIME ZONE,
        $3::VARCHAR,
        $4::BOOLEAN)
ON CONFLICT (account_id) DO UPDATE
SET
    valid_until = excluded.valid_until,
    last_status = CASE
      WHEN (account.valid_until < $5::TIMESTAMP WITH TIME ZONE)
      THEN $6::VARCHAR
      ELSE $7::VARCHAR
    END,
    force_publish = $8::BOOLEAN
RETURNING
    account.valid_until,
    account.last_status,
    (SELECT old_account.force_publish FROM account AS old_account WHERE id = id);

"""

```

**Reply by zzzeek** [reply, marked as answer]
oh, you can see one issue is the correlate requested there is not supported here.    SQLAlchemy does not support correlation from an INSERT, so the scalar subquery is not giving you what you want.   I'm not familiar with that syntax so you'd need to supply the ID directly

```py
).returning(
    Account.valid_until,
    Account.last_status,
    select(old_account.force_publish)
    .where(old_account.id == 1)
    .scalar_subquery()
    .label("old_force_publish"),
)

```

SQL:

```sql
INSERT INTO account (id, valid_until, last_status, force_publish) VALUES (%(id)s, %(valid_until)s, %(last_status)s, %(force_publish)s) ON CONFLICT (account_id) DO UPDATE SET valid_until = excluded.valid_until, last_status = CASE WHEN (account.valid_until < %(valid_until_1)s) THEN %(param_1)s ELSE %(param_2)s END, force_publish = %(param_3)s RETURNING account.valid_until, account.last_status, (SELECT account_1.force_publish 
FROM account AS account_1 
WHERE account_1.id = %(id_1)s) AS old_force_publish
```

**Reply by cpoppema** [reply]
You're so fast! :blush: Yes that looks exactly what I wanted! Supplying `account_id` as a parameter is not an issue. It runs, but for some reason `.where(old_account.id == 1)` is evaluated to `false` instead of an actual WHERE statement..

```
.. RETURNING 
    account.valid_until, 
    account.last_status,
    (SELECT account_1.force_publish 
    FROM account AS account_1 
    WHERE false) AS old_force_publish
```

**Reply by cpoppema** [reply]
> You're so fast! 😊 Yes that looks exactly what I wanted! Supplying `account_id` as a parameter is not an issue. It runs, but for some reason `.where(old_account.id == 1)` is evaluated to `false` instead of an actual WHERE statement..
> 
> ```
> .. RETURNING 
>     account.valid_until, 
>     account.last_status,
>     (SELECT account_1.force_publish 
>     FROM account AS account_1 
>     WHERE false) AS old_force_publish
> ```

Oh nevermind, I made a small typo!! It renders the correct sql now :sunglasses:
